### ISSUE TITLE
Simplify and standardize the PlantCV API

### ISSUE THREAD

**Issue by nfahlgren** [danforthcenter/plantcv#202]
### Description

This is more of a discussion than a specific task, we can define specific tasks based on discussion generated here.

For PlantCV3 I think we can do a couple of related things: simplify the API (reduce the number of inputs required for each function, merge or split some functions, etc.) and standardize how functions are named. I think it's reasonable to approach this with an eye towards maintaining backwards compatibility, at least for v3.

### Details

Some specific ideas, there could be a lot more added to this:

* Create a PlantCV class (let's call it `params` for now) that is automatically instantiated and contains methods that define tunable parameters (e.g. `device` counter, `debug` mode, etc.).
* Using the PlantCV `params` class, PlantCV functions have access to the `params` methods automatically, so these parameters no longer need to be passed directly to each function (in the example above we immediately remove at least two inputs to every function).
* The user would be able to override the default values in the `params` class to change the behavior of PlantCV.
* Add methods to `params` that would allow users to override default values like line and fill colors, thickness, and other plotting-type features. We would need to change these from hard-coded values in functions to variables that access the `params` class.
* Make new functions where necessary to change or replace the behavior of old functions. The old functions could be rebuilt to call the new version with settings that mimic legacy behavior for backwards compatibility. This could be done for issues like #49 
* In the course of updating any functions (where that makes sense), we can develop a standardized naming approach (and argument order) to make it easier to understand what does what.

### Completion Criteria

We can close this once we finalize a plan and create new, specific issues to move this idea forward (or not).

**Comment by ic** [comment]
The changes you propose would be, from my idiosyncratic use cases, very welcomed. Writing pipelines so far has two major issues:
* Boiler-plate: Many parameters must be repeated all along the pipeline. A language like Python allows so many improvements here.
* At times too thin a layer on top of OpenCV: A higher-lever API would make PlantCV even more useful to plant analysts, without too much prior knowledge of OpenCV. A lower API (the existing one?) could still remain for people with the brains and needs to refine further.

The thread mentions the boiler-plate---big win. I would be curious to hear about several practitioners how a "higher-level API" could be really more useful.

**Comment by nfahlgren** [comment]
@ic thanks for the feedback! I have a pull request in the pipeline for adding the `params` class discussed above, so I think we can start moving on the boiler-plate stuff pretty soon.

Your suggestion about the two API levels is interesting. I can see the value in it for different types of users. I'm trying to think of how it might be organized, happy to have any suggestions from you and others.

**Comment by ic** [comment]
The "multi-level" APIs is quite frequent practice.

A great, albeit complex, example at hand is Git itself. Most of us use the "user-friendly" API, called "porcelain". There is also the "plumbing" API, way more complex and powerful. The thing is that the porcelain is built on top of the plumbing.

Another example I see in my domain is Tensorflow. The API provides a lot of primitives, powerful and complex to put together. Over time, more and more "compound" API appear. These new APIs provide higher-level concepts, way easier to use as an ML user. Yet the primitive API remains for engineers to have full control over future (high-level) API opportunities, as well as performance tuning, etc.

Interestingly Tensorflow shows instances where a high-level API is logically equivallent to the composition of primitives, but completely reimplemented for better performance. A good example is [tf.nn.softmax_cross_entropy_with_logits](https://www.tensorflow.org/api_docs/python/tf/nn/softmax_cross_entropy_with_logits). In short, instead of just calling the softmax, cross entropy, and logit primitives, this younger API reimplements them for efficiency.
